### PR TITLE
Audit Logs

### PR DESCRIPTION
**Pull request by Snazzah**
This PR requires webhooks.
- [x] [Fix init bug](https://i-need.discord.cards/20b7db.png)
- [x] Test it after that bug fix

### Testing
- [x] Message entries work
- [x] Role entries work
- [x] User entries work
- [x] Guild entries work
- [x] Invite entries work
- [x] Emoji entries work
- [x] Channel entries work

### Other stuff to add
- [ ] Add `old_widget_channel` and `new_widget_channel` for changes with the `widget_channel_id` key
- [ ] Add `old_afk_channel` and `new_afk_channel` for changes with the `afk_channel_id` key
- [ ] Add `old_owner` and `new_owner` for changes with the `owner_id` key
- [x] Add Overwrite objects for changes with the `permission_overwrites` key **UNTESTED**

### CODE REVIEW

File: lib/discordrb/data.rb [hunk 1]
```diff
@@ -2970,6 +2975,274 @@ def process_voice_states(voice_states)
...
+    def get_target_type(action)
+      return :server if action < 10
+      return :channel if action < 20 
+      return :user if action < 30
+      return :role if action < 40
+      return :invite if action < 50
+      return :webhook if action < 60
+      return :emoji if action < 70
+      return :message if action < 80
+      :unknown
```
Comment (z64) [inline lib/discordrb/data.rb]: Can you refactor this into a `case`?

File: lib/discordrb/data.rb [hunk 2]
```diff
@@ -2970,6 +2975,274 @@ def process_voice_states(voice_states)
...
+        when :role
+          return @server.role(id)
+        when :invite
+          return @bot.invite(@data['changes'].find { |change| change['key'] == 'code' }.keys.delete_if { |k| k == 'key' }.first)
+        when :webhook
+          return 'TODO' # TODO
+        when :emoji
+          return @server.emoji[id]
+        end
+        nil
```
Comment (z64) [inline lib/discordrb/data.rb]: The `return`s and `nil` here should be omitted as `case` should implicitly return any of these conditions.

File: lib/discordrb/data.rb [hunk 3]
```diff
@@ -2970,6 +2975,274 @@ def process_voice_states(voice_states)
...
+        return @user unless @user.nil?
+        @user = @server.member(@data['user_id'].to_i) || @bot.user(@data['user_id'].to_i) || @logs.user(@data['user_id'].to_i)
+      end
+      alias_method :executor, :user
+
+      # @return [Channel, nil] the amount of messages deleted. Is not nil if the action is `:message_delete`.
+      def channel
+        return nil unless @channel_id
+        return @channel unless @channel.nil?
+        @channel = @bot.channel(@channel_id, @server)
```
Comment (z64) [inline lib/discordrb/data.rb]: The above two lines can be simplified:

`@channel ||= @bot.channel(@channel_id, @server)`

File: lib/discordrb/data.rb [hunk 4]
```diff
@@ -2970,6 +2975,274 @@ def process_voice_states(voice_states)
...
+      def target
+        return @target if @target_cached
+        @target_cached = true
+        @target = process_target(@data['target_id'], @target_type)
+      end
+
+      # @return [Member, User] the user that is executing this action. Can be a User object if the user no longer exists in the server.
+      def user
+        return @user unless @user.nil?
+        @user = @server.member(@data['user_id'].to_i) || @bot.user(@data['user_id'].to_i) || @logs.user(@data['user_id'].to_i)
```
Comment (z64) [inline lib/discordrb/data.rb]: The above two lines can be simplified:

`@user ||= @server.member #...`

File: lib/discordrb/data.rb [hunk 5]
```diff
@@ -2970,6 +2975,274 @@ def process_voice_states(voice_states)
...
+        @user = nil
+        @action = Actions[data['action_type']]
+        @action_type = @logs.get_action_type(data['action_type'])
+        @target_type = @logs.get_target_type(data['action_type'])
+        @target_cached = false
+        @count = nil
+        @channel = nil
+        @channel_id = nil
+        @days = nil
+        @members_removed = nil
```
Comment (z64) [inline lib/discordrb/data.rb]: It's pointless setting any of the above a la `@attr = nil`.

File: lib/discordrb/data.rb [hunk 6]
```diff
@@ -2970,6 +2975,274 @@ def process_voice_states(voice_states)
...
+        @target_cached = true
+        @target = process_target(@data['target_id'], @target_type)
+      end
+
+      # @return [Member, User] the user that is executing this action. Can be a User object if the user no longer exists in the server.
+      def user
+        return @user unless @user.nil?
+        @user = @server.member(@data['user_id'].to_i) || @bot.user(@data['user_id'].to_i) || @logs.user(@data['user_id'].to_i)
+      end
+      alias_method :executor, :user
```
Comment (z64) [inline lib/discordrb/data.rb]: I think `executer` isn't the best name for this alias. I might suggest `author` to be consistent with similar methods in the lib that may return a `User` or other `User`-like type.

File: lib/discordrb/data.rb [hunk 7]
```diff
@@ -2970,6 +2975,274 @@ def process_voice_states(voice_states)
...
+        @channel_id = data['options']['channel'].to_i unless data['options']['channel'].nil?
+        @days = data['options']['delete_member_days'].to_i unless data['options']['delete_member_days'].nil?
+        @members_removed = data['options']['members_removed'].to_i unless data['options']['members_removed'].nil?
+      end
+
+      # @return [Server, Channel, Member, User, Role, Invite, Webhook, Emoji, nil] the target being preformed on.
+      def target
+        return @target if @target_cached
+        @target_cached = true
+        @target = process_target(@data['target_id'], @target_type)
```
Comment (z64) [inline lib/discordrb/data.rb]: I'm not sure why `@target_cached` is necessary.

`@target ||= process_target(@data['target_id'], @target_type)`, then any other place that expects `@target_cached` (which, i can't find anywhere that actually uses this attribute) can just check that `@target` is truthy / not `nil`.

File: lib/discordrb/data.rb [hunk 8]
```diff
@@ -2970,6 +2975,274 @@ def process_voice_states(voice_states)
...
+        return @channel unless @channel.nil?
+        @channel = @bot.channel(@channel_id, @server)
+      end
+
+      # @!visibility private
+      def process_target(id, type)
+        id = id.resolve_id
+        case type
+        when :server
+          return @server # Since it won't be anything else
```
Comment (z64) [inline lib/discordrb/data.rb]: This comment doesn't really tell the reader anything.

File: lib/discordrb/data.rb [hunk 9]
```diff
@@ -2970,6 +2975,274 @@ def process_voice_states(voice_states)
...
+
+    # Checks whether a member is cached
+    # @note For internal use only
+    # @!visibility private
+    def user_cached?(user_id)
+      @users.include?(user_id)
+    end
+
+    def process_users(users)
+      return unless users
```
Comment (z64) [inline lib/discordrb/data.rb]: This check is pointless. And, this method isn't even called anywhere?

File: lib/discordrb/data.rb [hunk 10]
```diff
@@ -2970,6 +2975,274 @@ def process_voice_states(voice_states)
...
+      # @return [Symbol] the action that was preformed.
+      attr_reader :action
+
+      # @return [Symbol] the type action that was preformed. (:create, :delete, :update, :unknown)
+      attr_reader :action_type
+
+      # @return [Symbol] the type of target being preformed on. (:server, :channel, :user, :role, :invite, :webhook, :emoji, :unknown)
+      attr_reader :target_type
+
+      # @return [Integer, nil] the amount of messages deleted. Is not nil if the action is `:message_delete`.
```
Comment (z64) [inline lib/discordrb/data.rb]: Kind of awkward wording.
```
@return [Integer, nil] the amount of messages deleted if the `:action_type` is `:message_delete`.
```

File: lib/discordrb/data.rb [hunk 11]
```diff
@@ -2970,6 +2975,274 @@ def process_voice_states(voice_states)
...
+      attr_reader :count
+      alias_method :amount, :count
+
+      # @return [Integer, nil] the amount of days the members were inactive for. Is not nil if the action is `:member_prune`.
```
Comment (z64) [inline lib/discordrb/data.rb]: "Is not nil"

Same as above

File: lib/discordrb/data.rb [hunk 12]
```diff
@@ -2970,6 +2975,274 @@ def process_voice_states(voice_states)
...
+      attr_reader :days
+
+      # @return [Integer, nil] the amount of members removed. Is not nil if the action is `:member_prune`.
```
Comment (z64) [inline lib/discordrb/data.rb]: "Is not nil"

Same as above

File: lib/discordrb/data.rb [hunk 13]
```diff
@@ -2970,6 +2975,274 @@ def process_voice_states(voice_states)
...
+      attr_reader :members_removed
+
+      # @return [Hash<String => Change>, nil] the changes from this log, listing the key as the key changed. Is nil if the action is either `:message_delete` or `:member_prune`.
```
Comment (z64) [inline lib/discordrb/data.rb]: Just tweak the wording on this one:
```
This is `nil` if the action is either `:message_delete` or `:member_prune`.
```

File: lib/discordrb/data.rb [hunk 14]
```diff
@@ -2970,6 +2975,274 @@ def process_voice_states(voice_states)
...
+    class Change
+
+      # @return [String] the key that was changed.
+      # @note You should check with the Discord API Documentation on what key gives out what value.
+      attr_reader :key
+
+      # @return [String, Integer, true, false, Permissions, nil] the value that was changed from.
+      attr_reader :old
+      alias_method :old_value, :old
+      alias_method :old_val, :old
```
Comment (z64) [inline lib/discordrb/data.rb]: I don't think these aliases are useful, particularly `:old_val`. I would make the attribute itself `:old_value` and alias `:old` if you want, but nothing more.

File: lib/discordrb/data.rb [hunk 15]
```diff
@@ -2970,6 +2975,274 @@ def process_voice_states(voice_states)
...
+
+      # @return [String, Integer, true, false, Permissions, nil] the value that was changed to.
+      attr_reader :new
+      alias_method :new_value, :new
+      alias_method :new_val, :new
```
Comment (z64) [inline lib/discordrb/data.rb]: Same as above for these aliases

File: lib/discordrb/data.rb [hunk 16]
```diff
@@ -2970,6 +2975,274 @@ def process_voice_states(voice_states)
...
+      @entries.first
+    end
+    alias_method :first, :latest
+
+    # Gets a member in the audit logs data based on user ID
+    # @note This only uses data given by the audit logs request
+    # @param id [Integer] The user ID to look for
+    def user(id)
+      id = id.resolve_id
+      return @users[id] if user_cached?(id)
```
Comment (z64) [inline lib/discordrb/data.rb]: I'm not sure why this check and method (`user_cached?`) is necessary. Just return `@user[id]`.

File: lib/discordrb/data.rb [hunk 17]
```diff
@@ -2970,6 +2975,274 @@ def process_voice_states(voice_states)
...
+        @action_type = @logs.get_action_type(data['action_type'])
+        @target_type = @logs.get_target_type(data['action_type'])
+        @target_cached = false
+        @count = nil
+        @channel = nil
+        @channel_id = nil
+        @days = nil
+        @members_removed = nil
+        @changes = @action == :message_delete || @action == :member_prune ? nil : {}
+        process_changes(data['changes'])
```
Comment (z64) [inline lib/discordrb/data.rb]: This method `process_changes` below internally has a guard.

`return unless changes`

I would rather the guard be here instead and not inside the method

`process_changes(data['changes']) unless data['changes']`

**Comment by z64** [comment]
A more general note:

- Please run `rubocop` locally before committing.
- Use less terrible commit messages. I imagine it's going to be squashed, anyway.

File: lib/discordrb/data.rb (hunk 7 above)
Comment (SnazzyPine25) [inline lib/discordrb/data.rb]: It could be null if a unhandled entry comes in.

File: lib/discordrb/data.rb (hunk 6 above)
Comment (SnazzyPine25) [inline lib/discordrb/data.rb]: I'll add a alias for `author`

File: lib/discordrb/data.rb (hunk 2 above)
Comment (SnazzyPine25) [inline lib/discordrb/data.rb]: Rubocop yells at me if I do that.

File: lib/discordrb/data.rb (hunk 14 above)
Comment (SnazzyPine25) [inline lib/discordrb/data.rb]: mk

File: lib/discordrb/data.rb [hunk 18]
```diff
@@ -2970,6 +2975,273 @@ def process_voice_states(voice_states)
...
+      # @return [Symbol] the action that was preformed.
+      attr_reader :action
+
+      # @return [Symbol] the type action that was preformed. (:create, :delete, :update, :unknown)
+      attr_reader :action_type
+
+      # @return [Symbol] the type of target being preformed on. (:server, :channel, :user, :role, :invite, :webhook, :emoji, :unknown)
+      attr_reader :target_type
+
+      # @return [Integer, nil] the amount of messages deleted. Won't be if the action is `:message_delete`.
```
Comment (z64) [inline lib/discordrb/data.rb]: I still think this form is less than ideal. You can refer to my previous comment.

However, sticking with what you wrote, you're missing word `nil` here.

File: lib/discordrb/data.rb [hunk 19]
```diff
@@ -2970,6 +2975,273 @@ def process_voice_states(voice_states)
...
+        @user ||= @server.member(@data['user_id'].to_i) || @bot.user(@data['user_id'].to_i) || @logs.user(@data['user_id'].to_i)
+      end
+      alias_method :executor, :user
+      alias_method :author, :user
+
+      # @return [Channel, nil] the amount of messages deleted. Won't be nil if the action is `:message_delete`.
+      def channel
+        return nil unless @channel_id
+        return @channel unless @channel.nil?
+        @channel = @bot.channel(@channel_id, @server)
```
Comment (z64) [inline lib/discordrb/data.rb]: Please refactor like the others.

`@channel ||= @bot.channel(@channel_id, @server)`

File: lib/discordrb/data.rb [hunk 20]
```diff
@@ -2970,6 +2975,273 @@ def process_voice_states(voice_states)
...
+
+      # @!visibility private
+      def initialize(data)
+        @key = data['key']
+        @old = data['old_value']
+        @new = data['new_value']
+        evaluate_permissions
+      end
+
+      def evaluate_permissions
```
Comment (z64) [inline lib/discordrb/data.rb]: I don't really see why this needs to be its own (at least public) method - no one would use this. I would just append this to the body of `initialize`.

File: lib/discordrb/data.rb [hunk 21]
```diff
@@ -2970,6 +2975,268 @@ def process_voice_states(voice_states)
...
+      @server = server
+      @users = {}
+      @entries = data['audit_log_entries'].map { |entry| Entry.new(self, @server, @bot, entry) }
+    end
+
+    # An entry in a server's audit logs.
+    class Entry
+      include IDObject
+
+      # @return [Symbol] the action that was preformed.
```
Comment (meew0) [inline lib/discordrb/data.rb]: `performed`

File: lib/discordrb/data.rb [hunk 22]
```diff
@@ -2970,6 +2975,268 @@ def process_voice_states(voice_states)
...
+        when :channel
+          @bot.channel(id, @server)
+        when :user, :message
+          @server.member(id) || @bot.user(id) || @logs.user(id)
+        when :role
+          @server.role(id)
+        when :invite
+          @bot.invite(@data['changes'].find { |change| change['key'] == 'code' }.values.delete_if { |v| v == 'code' }.first)
+        when :webhook
+          'TODO' # TODO
```
Comment (meew0) [inline lib/discordrb/data.rb]: Please implement this

File: lib/discordrb/data.rb [hunk 23]
```diff
@@ -2970,6 +2975,268 @@ def process_voice_states(voice_states)
...
+      # @return [Symbol] the action that was preformed.
+      attr_reader :action
+
+      # @return [Symbol] the type action that was preformed. (:create, :delete, :update, :unknown)
+      attr_reader :action_type
+
+      # @return [Symbol] the type of target being preformed on. (:server, :channel, :user, :role, :invite, :webhook, :emoji, :unknown)
+      attr_reader :target_type
+
+      # @return [Integer, nil] the amount of messages deleted. Won't be nil if the action is `:message_delete`.
```
Comment (meew0) [inline lib/discordrb/data.rb]: I'd change "Won't be nil" to "Only present" in all cases, it appears more clear to me

File: lib/discordrb/data.rb [hunk 24]
```diff
@@ -2970,6 +2975,268 @@ def process_voice_states(voice_states)
...
+    end
+
+    # An entry in a server's audit logs.
+    class Entry
+      include IDObject
+
+      # @return [Symbol] the action that was preformed.
+      attr_reader :action
+
+      # @return [Symbol] the type action that was preformed. (:create, :delete, :update, :unknown)
```
Comment (meew0) [inline lib/discordrb/data.rb]: `performed`

File: lib/discordrb/data.rb [hunk 25]
```diff
@@ -2970,6 +2975,268 @@ def process_voice_states(voice_states)
...
+      attr_reader :action_type
+
+      # @return [Symbol] the type of target being preformed on. (:server, :channel, :user, :role, :invite, :webhook, :emoji, :unknown)
```
Comment (meew0) [inline lib/discordrb/data.rb]: `performed`

File: lib/discordrb/data.rb [hunk 26]
```diff
@@ -2970,6 +2975,268 @@ def process_voice_states(voice_states)
...
+      # @return [Integer, nil] the amount of members removed. Won't be nil if the action is `:member_prune`.
+      attr_reader :members_removed
+
+      # @return [Hash<String => Change>, RoleChange, nil] the changes from this log, listing the key as the key changed. Will be a RoleChange object if the action is `:member_role_update`. Will be nil if the action is either `:message_delete` or `:member_prune`.
+      attr_reader :changes
+
+      # @!visibility private
+      def initialize(logs, server, bot, data)
+        @bot = bot
+        @id = data['id']
```
Comment (meew0) [inline lib/discordrb/data.rb]: This should be an integer

File: lib/discordrb/data.rb [hunk 27]
```diff
@@ -2970,6 +2975,268 @@ def process_voice_states(voice_states)
...
+        @target_type = @logs.get_target_type(data['action_type'])
+        @target_cached = false
+        @changes = {} unless @action == :message_delete || @action == :member_prune || @action == :member_role_update
+        @changes = RoleChange.new(data['changes'][0], @server) if @action == :member_role_update
+        process_changes(data['changes']) unless @action == :member_role_update
+        return unless data.include?('options')
+        @count = data['options']['count'].to_i unless data['options']['count'].nil?
+        @channel_id = data['options']['channel'].to_i unless data['options']['channel'].nil?
+        @days = data['options']['delete_member_days'].to_i unless data['options']['delete_member_days'].nil?
+        @members_removed = data['options']['members_removed'].to_i unless data['options']['members_removed'].nil?
```
Comment (meew0) [inline lib/discordrb/data.rb]: Can you add some newlines and comments to the code in this method to make it more clear what everything does?

File: lib/discordrb/data.rb [hunk 28]
```diff
@@ -2970,6 +2975,268 @@ def process_voice_states(voice_states)
...
+
+    def process_users(users)
+      users.each do |element|
+        user = User.new(element, self, @bot)
+        @users[user.id] = user
+      end
+    end
+
+    # @!visibility private
+    def get_target_type(action)
```
Comment (meew0) [inline lib/discordrb/data.rb]: This method can be static, also please rename it to `target_type_for` or something (starting method names with `get` isn't good Ruby practice)

File: lib/discordrb/data.rb [hunk 29]
```diff
@@ -2970,6 +2975,268 @@ def process_voice_states(voice_states)
...
+      return :role if action < 40
+      return :invite if action < 50
+      return :webhook if action < 60
+      return :emoji if action < 70
+      return :message if action < 80
+      :unknown
+    end
+
+    # @!visibility private
+    def get_action_type(action)
```
Comment (meew0) [inline lib/discordrb/data.rb]: This method can be static, also please rename it to `action_type_for` or something (starting method names with `get` isn't good Ruby practice)

File: lib/discordrb/data.rb [hunk 30]
```diff
@@ -2970,6 +2975,268 @@ def process_voice_states(voice_states)
...
+        @changes = RoleChange.new(data['changes'][0], @server) if @action == :member_role_update
+        process_changes(data['changes']) unless @action == :member_role_update
+        return unless data.include?('options')
+        @count = data['options']['count'].to_i unless data['options']['count'].nil?
+        @channel_id = data['options']['channel'].to_i unless data['options']['channel'].nil?
+        @days = data['options']['delete_member_days'].to_i unless data['options']['delete_member_days'].nil?
+        @members_removed = data['options']['members_removed'].to_i unless data['options']['members_removed'].nil?
+      end
+
+      # @return [Server, Channel, Member, User, Role, Invite, Webhook, Emoji, nil] the target being preformed on.
```
Comment (meew0) [inline lib/discordrb/data.rb]: `performed`

File: lib/discordrb/data.rb (hunk 1 above)
Comment (SnazzyPine25) [inline lib/discordrb/data.rb]: I think i wont be able to compare it to `<`.

Comment (Daniel-Worrall) [inline lib/discordrb/data.rb]: Ruby case supports ranges.

File: lib/discordrb/data.rb [hunk 31]
```diff
@@ -3114,6 +3125,279 @@ def process_voice_states(voice_states)
...
+    # @return [Array<User>] the users included in the audit logs.
+    attr_reader :users
+
+    # @return [Array<Entry>] the entries listed in the audit logs.
+    attr_reader :entries
+
+    # @!visibility private
+    def initialize(server, bot, data)
+      @bot = bot
+      @id = data['id'].resolve_id
```
Comment (meew0) [inline lib/discordrb/data.rb]: Why is this here? It causes audit log creation to fail at all times, because the response doesn't have an `id` property. **(TEST YOUR CODE)**

**Comment by SnazzyPine25** [comment]
Sorry, I mean't to only have that for entries

File: lib/discordrb/api/server.rb [hunk 32]
```diff
@@ -320,6 +320,17 @@ def invites(token, server_id)
     )
   end
 
+  # Gets a server's audit logs
+  def audit_logs(token, server_id, limit, userid = nil, actiontype = nil)
```
Comment (z64) [inline lib/discordrb/api/server.rb]: This was discussed, but just so it isn't forgotten:

This is missing the `before` querystring param.

[Reference the WIP documentation](https://github.com/hammerandchisel/discord-api-docs/blob/fb5549435c9c8b10a0ade96d68bea3921ea496d5/docs/resources/Audit_Log.md#get-guild-audit-log--get-guildsguildiddocs_guildguild-objectaudit-log)

**Comment by Daniel-Worrall** [comment]
This was asked about in the discord channel. Remember you have 3 failing rubocop checks that need fixing before this can be merged.

File: lib/discordrb/data.rb [hunk 33]
```diff
@@ -3114,6 +3127,258 @@ def process_voice_states(voice_states)
...
+        @channel_id = data['options']['channel'].to_i unless data['options']['channel'].nil?
+        @days = data['options']['delete_member_days'].to_i unless data['options']['delete_member_days'].nil?
+        @members_removed = data['options']['members_removed'].to_i unless data['options']['members_removed'].nil?
+      end
+
+      # @return [Server, Channel, Member, User, Role, Invite, Webhook, Emoji, nil] the target being performed on.
+      def target
+        return @target if @target_cached
+        @target_cached = true
+        @target = process_target(@data['target_id'], @target_type)
```
Comment (LikeLakers2) [inline lib/discordrb/data.rb]: I'd personally recommend we bring the `@target_cached = true` down below the `@target =` line -- if something goes wrong in `process_target` just due to random luck and causes an exception, it won't leave us with always getting `nil` in the future.

**Comment by LikeLakers2** [comment]
Additionally, quite a bit of code I see here seems to rely on hardcoded patterns -- for example, `action_type_for` relies on a pre-defined list of keys to figure out whether to assign `:create`, `:update`, or `:delete`. While I understand that we cannot account for everything, I'm wondering if there is some better solution that could keep this working without needing to update every time a new audit log type is added.

It may be useful to assign the entire data hash we get to its own variable, especially if we can't figure out everything in the audit log ourselves through code. This would allow any audit logs that *we* can't figure out to still be figured out by the end developer.

File: lib/discordrb/data.rb (hunk 7 above)
Comment (Daniel-Worrall) [inline lib/discordrb/data.rb]: That's exactly what the `||=` operator handles. If `@target` is `nil`, it'll call the `process_target` as needed. If it's assigned (because it has already called `process_target`) then it will just return from cache.

File: lib/discordrb/data.rb [hunk 34]
```diff
@@ -3114,6 +3127,259 @@ def process_voice_states(voice_states)
...
+        @days = data['options']['delete_member_days'].to_i unless data['options']['delete_member_days'].nil?
+        @members_removed = data['options']['members_removed'].to_i unless data['options']['members_removed'].nil?
+      end
+
+      # @return [Server, Channel, Member, User, Role, Invite, Webhook, Emoji, nil] the target being performed on.
+      def target
+        return @target if @target_cached
+        @target = process_target(@data['target_id'], @target_type)
+        @target_cached = true
+        return @target
```
Comment (Daniel-Worrall) [inline lib/discordrb/data.rb]: All this should just be 
```
 @target ||= process_target(@data['target_id'], @target_type)
```
There's no need to introduce an extra variable to handle checking whether the target has been cached as standard Rubyisms are to just check if it's nil or not, which is exactly what's happening.

This also removes the redundant returns. Make sure you remove the `@target_cached` initialise too.

File: lib/discordrb/api/server.rb [hunk 35]
```diff
@@ -320,6 +320,17 @@ def invites(token, server_id)
     )
   end
 
+  # Gets a server's audit logs
```
Comment (z64) [inline lib/discordrb/api/server.rb]: Please add a URL to the documentation for this route.

https://discordapp.com/developers/docs/resources/audit-log#get-guild-audit-log

File: lib/discordrb/data.rb [hunk 36]
```diff
@@ -2653,6 +2653,19 @@ def integrations
...
+    # @param action [Symbol] The action to only include.
+    # @param user [User, #resolve_id] The user to filter entries to.
+    # @param limit [Integer] The amount of entries to limit it to.
+    # @param before [Entry, #resolve_id] The entry to use to not include all entries after it.
+    # @return [AuditLogs] the server's audit logs.
+    def audit_logs(action: nil, user: nil, limit: 50, before: nil)
+      raise 'Invalid audit log action!' if action && AuditLogs::Actions.key(action).nil?
+      action = AuditLogs::Actions.key(action)
+      user = user.resolve_id if user.respond_to? :resolve_id
+      before = before.resolve_id if before.respond_to? :resolve_id
```
Comment (z64) [inline lib/discordrb/data.rb]: Please remove these guards. The arguments should respond to `resolve_id`, if they don't, they are likely passing in a bad object which will generate a `400 Bad Request`.

File: lib/discordrb/data.rb [hunk 37]
```diff
@@ -3164,6 +3164,9 @@ class AuditLogs
     # @return [Array<User>] the users included in the audit logs.
     attr_reader :users
 
+    # @return [Array<Webhook>] the users included in the audit logs.
```
Comment (z64) [inline lib/discordrb/data.rb]: image

(The type `Array<Webhook>` is also incorrect)

File: lib/discordrb/data.rb [hunk 38]
```diff
@@ -3333,23 +3340,40 @@ def latest
...
+    # Gets a user in the audit logs data based on user ID
     # @note This only uses data given by the audit logs request
     # @param id [Integer] The user ID to look for
     def user(id)
       @users[id.resolve_id]
     end
 
+    # Gets a webhook in the audit logs data based on webhook ID
+    # @note This only uses data given by the audit logs request
+    # @param id [Integer] The webhook ID to look for
```
Comment (z64) [inline lib/discordrb/data.rb]: `@param id [Integer, #resolve_id]`

File: lib/discordrb/data.rb [hunk 39]
```diff
@@ -3333,23 +3340,40 @@ def latest
...
+        user = User.new(element, @bot)
         @users[user.id] = user
       end
     end
 
+    # Process webhook objects given by the request
+    # @note For internal use only
+    # @!visibility private
+    def process_webhooks(webhooks)
+      users.each do |element|
```
Comment (z64) [inline lib/discordrb/data.rb]: image

File: lib/discordrb/data.rb [hunk 40]
```diff
@@ -3164,6 +3164,9 @@ class AuditLogs
     # @return [Array<User>] the users included in the audit logs.
```
Comment (z64) [inline lib/discordrb/data.rb]: The type `Array<User>` is incorrect.

File: lib/discordrb/data.rb (hunk 33 above)
Comment (SnazzyPine25) [inline lib/discordrb/data.rb]: If something comes up as deleted and cant get retrieved, it'll be nil.

Comment (LikeLakers2) [reply lib/discordrb/data.rb]: Yes, we know, but why would we assume this cached successfully *before* processing the target? It doesn't make sense.

Also, as pointed out in the review that doesn't show up in changes anymore, why do you even need `@target_cached`? Just do `@target ||= process_target(@data['target_id'], @target_type)`.

File: lib/discordrb/api/server.rb [hunk 41]
```diff
@@ -296,6 +296,31 @@ def remove_member_role(token, server_id, user_id, role_id, reason = nil)
     )
   end
 
+  # Adds a single role to a member
+  # https://discordapp.com/developers/docs/resources/guild#add-guild-member-role
+  def add_member_role(token, server_id, user_id, role_id)
```
Comment (z64) [inline lib/discordrb/api/server.rb]: As mentioned in chat, remove these as since you rebased, they're (this method and `remove_member_role` below) in this file twice now.

File: lib/discordrb/api/server.rb [hunk 42]
```diff
@@ -460,4 +497,15 @@ def webhooks(token, server_id)
       Authorization: token
     )
   end
+
+  # Available voice regions for this server
+  def regions(token, server_id)
```
Comment (z64) [inline lib/discordrb/api/server.rb]: ditto

File: lib/discordrb/data.rb [hunk 43]
```diff
@@ -3324,6 +3337,319 @@ def update_webhook(new_data)
...
+      52 => :webhook_delete,
+      60 => :emoji_create,
+      61 => :emoji_update,
+      62 => :emoji_delete,
+      # 70
+      # 71
+      72 => :message_delete
+    }.freeze
+
+    # @return [Hash<String, User>] the users included in the audit logs.
```
Comment (z64) [inline lib/discordrb/data.rb]: If you see [this commit](https://github.com/meew0/discordrb/pull/409/commits/064f024e864a73ccddd90863396160bd0cf937eb), there was only a few places where we used this style of `Hash` docstring, so I changed it to use the "hash rocket" style to be consistent with the majority. Would you mind doing this here?

File: lib/discordrb/data.rb [hunk 44]
```diff
@@ -3324,6 +3337,319 @@ def update_webhook(new_data)
...
+    attr_reader :users
+
+    # @return [Hash<String, Webhook>] the webhooks included in the audit logs.
```
Comment (z64) [inline lib/discordrb/data.rb]: ditto

**Comment by z64** [comment]
I've merged this as 6f479221297c1b7be4aa79ba4b46e926098ebca4 with some fixes applied. I've tested this implementation as best I can, but I don't really have a use case for audit logs so I'm unsure if the API could be made better.

Since this PR has been opened there hasn't been many requests for audit log features, so while I can say this works as presented, if someone needs audit logs they're welcome to offer improvements.

Thanks!
